File: examples/sync.rs
#[macro_use]
extern crate serde_derive;

#[calcite::deno_op]
fn array_buffer_example(mut arr: calcite::ArrayBuffer<i32>, num: i32) {
    for (pos, val) in arr.as_mut_slice().iter_mut().enumerate() {
        *val = num + pos as i32;
    }
}

#[derive(Deserialize, Debug)]
struct Struct1<'a> {
    a: i32,
    b: &'a str,
}

#[calcite::deno_op]
fn struct_example(s1: Struct1) {
    println!("Got argument {:?}", s1);
}

#[calcite::deno_op]
fn return_example(a: (i32, &str)) -> String {
    format!("I got {:?} as argument", a)
}

#[calcite::deno_op]
fn multiple_arguments_example(x: f32, y: (&str, &str)) {
    println!("Got arguments x:{} y:{:?}", x, y);
}

#[calcite::deno_op]
fn return_buffer() -> calcite::ReturnBuffer {
    calcite::ReturnBuffer::from_bytes(vec![5, 7, 1, 3])
}

calcite::export!(
    array_buffer_example,
    struct_example,
    return_example,
    multiple_arguments_example,
    return_buffer
);

File: examples/sync.ts
import {
  loadPlugin,
  importFromPlugin,
} from "https://deno.land/x/calcite@2.0/calcite.ts";
await loadPlugin("sync", "file://target/debug/examples/");

const array_buffer_example = importFromPlugin("array_buffer_example") as (
  arr: Int32Array,
  num: number,
) => void;
const struct_example = importFromPlugin("struct_example") as (
  s1: { a: number; b: string },
) => void;
const return_example = importFromPlugin("return_example") as (
  a: [number, string],
) => string;
const multiple_arguments_example = importFromPlugin(
  "multiple_arguments_example",
) as (x: number, y: [string, string]) => void;
const return_buffer = importFromPlugin(
  "return_buffer",
  { returnRawBuffer: true },
) as () => Uint8Array;

let a = new Int32Array(5);
array_buffer_example(a, 8);
console.log(`The value of a is ${a}`);

struct_example({ a: 75, b: "Hey!" });
console.log(
  `The result of return_example is "${return_example([98, "Bye!"])}"`,
);

multiple_arguments_example(8.9, ["abc", "xyz"]);
console.log(return_buffer());
